File: Anul 2/Semestrul 2/MPP/labs/Tema 5-proiect Server Csharp-Client Java/proiectSCcsharp/angajatNetworking/ServicesRpcProxy.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization.Formatters.Binary;
using System.Runtime.Serialization;
using System.Threading;
using proiectSCcsharp.services;
using proiectSCcsharp.domain;
using log4net;

namespace proiectSCcsharp.networking
{

    public class ServicesRpcProxy : IService
    {
        private static readonly ILog Logger = LogManager.GetLogger("ServicesRpcProxy");
        private readonly string host;
        private readonly int port;
        private IObserver angajatObserver;
        private NetworkStream stream;
        private IFormatter formatter;
        private TcpClient connection;
        private readonly Queue<Response> responses;
        private volatile bool finished;
        private EventWaitHandle waitHandle;

        public ServicesRpcProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<Response>();
        }

        private void CloseConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                waitHandle.Close();
                angajatObserver = null;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void SendRequest(Request request)
        {
            try
            {
                formatter.Serialize(stream, request);
                stream.Flush();
            }
            catch (Exception e)
            {
                throw new MyException("Error sending object " + e);
            }
        }

        private Response ReadResponse()
        {
            Response response = null;
            try
            {
                waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
            return response;
        }

        private void InitializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                waitHandle = new AutoResetEvent(false);
                StartReader();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private void StartReader()
        {
            var tw = new Thread(Run);
            tw.Start();
        }

        private void HandleUpdate(Response response)
        {
            if (response.Type != ResponseType.SOLD_BILET) return;
            var bilet = (Bilet)response.Data;
            Logger.Info("Bilet sold " + bilet);
            try
            {
                angajatObserver.BiletAdded(bilet);
            }
            catch (MyException e)
            {
                Logger.Error("Error handle update: " + e);
            }
        }

        private static bool IsUpdate(Response response)
        {
            return response.Type == ResponseType.SOLD_BILET;
        }

        public Angajat login(Angajat angajat, IObserver angajatObserver)
        {
            InitializeConnection();
            var req = new Request.Builder().Type(RequestType.LOGIN).Data(angajat).Build();
            SendRequest(req);
            var response = ReadResponse();
            switch (response.Type)
            {
                case ResponseType.OK:
                    this.angajatObserver = angajatObserver;
                    Logger.Info("Logged in");
                    return (Angajat)response.Data;
                case ResponseType.ERROR:
                    {
                        Logger.Error("Error logging in" + response.Data.ToString());
                        var err = response.Data.ToString();
                        CloseConnection();
                        throw new MyException(err);
                    }
                default:
                    return null;
            }
        }

        public void logout(Angajat angajat)
        {
            var req = new Request.Builder().Type(RequestType.LOGOUT).Data(angajat).Build();
            SendRequest(req);
            var response = ReadResponse();
            CloseConnection();
            if (response.Type == ResponseType.ERROR)
            {
                Logger.Error("Error logging out" + response.Data);
                var err = response.Data.ToString();
                throw new MyException(err);
            }
            Logger.Info("Logged out");
        }

        public void addBilet(Bilet bilet)
        {
            var req = new Request.Builder().Type(RequestType.ADD_BILET).Data(bilet).Build();
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                Logger.Error("Error adding bilet" + response.Data.ToString());
                var err = response.Data.ToString();
                throw new MyException(err);
            }
            Logger.Info("Bilet added");
            Console.WriteLine("Bilet added");
        }

        public Spectacol getSpectacol(long id)
        {
            var req = new Request.Builder().Type(RequestType.GET_SPECTACOL).Data(id).Build();
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                Logger.Error("Error finding spectacol" + response.Data.ToString());
                var err = response.Data.ToString();
                throw new MyException(err);
            }
            Logger.Info("Found spectacol");
            return (Spectacol)response.Data;
        }

        public List<Spectacol> getAllSpectacole()
        {
            var req = new Request.Builder().Type(RequestType.GET_ALL_SPECTACOLE).Data(null).Build();
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == ResponseType.ERROR)
            {
                Logger.Error("Error getting spectacole" + response.Data.ToString());
                Console.WriteLine("Error getting spectacole" + response.Data.ToString());
                var err = response.Data.ToString();
                throw new MyException(err);
            }
            Logger.Info("Got spectacole");
            Console.WriteLine("Got spectacole");
            return (List<Spectacol>)response.Data;
        }
        protected virtual void Run()
        {
            while (!finished)
            {
                try
                {
                    var response = formatter.Deserialize(stream);
                    Console.WriteLine("response received " + response);
                    if (IsUpdate((Response)response))
                    {
                        HandleUpdate((Response)response);
                    }
                    else
                    {
                        lock (responses)
                        {
                            responses.Enqueue((Response)response);
                        }
                        waitHandle.Set();
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine("Reading error " + e);
                }
            }
        }
    }

}

File: Anul 2/Semestrul 2/MPP/labs/Tema 5-proiect Server Csharp-Client Java/proiectServerClientJava/Networking/src/main/java/proiectServerClientJava/network/rpcprotocol/ServicesRpcProxy.java
package proiectServerClientJava.network.rpcprotocol;

import proiectServerClientJava.domain.Angajat;
import proiectServerClientJava.domain.Spectacol;
import proiectServerClientJava.domain.SpectacolDTO;
import proiectServerClientJava.domain.Bilet;
import proiectServerClientJava.service.IService;
import proiectServerClientJava.service.IObserver;
import proiectServerClientJava.service.MyException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.net.SocketException;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ServicesRpcProxy implements IService {
    private final String host;
    private final int port;

    private IObserver employeeObserver;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;
    private static final Logger logger = LogManager.getLogger();

    private final BlockingQueue<Response> qresponses;
    private volatile boolean finished;

    public ServicesRpcProxy(String host, int port) {
        logger.info("Creating proxy");
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingQueue<>();
    }

    private void closeConnection() {
        finished = true;
        try {
            input.close();
            output.close();
            connection.close();
            employeeObserver = null;
            logger.info("Closed connection");
        } catch (IOException e) {
            logger.error("Error closing connection: " + e);
        }
    }

    private void sendRequest(Request request) throws MyException {
        try {
            output.writeObject(request);
            output.flush();
            logger.info("Request sent: " + request);
        } catch (IOException e) {
            logger.error("Error sending object " + e);
            throw new MyException("Error sending object " + e);
        }
    }

    private Response readResponse() throws MyException {
        Response response = null;
        try {
            response = qresponses.take();
            logger.info("Response received: " + response);
        } catch (InterruptedException e) {
            logger.error("Reading response error " + e);
            throw new MyException("Reading response error " + e);
        }
        return response;
    }

    private void initializeConnection() throws MyException {
        try {
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
            logger.info("Connection initialized");
        } catch (IOException e) {
            logger.error("Error connecting to server " + e);
            throw new MyException("Error connecting to server " + e);
        }
    }

    private void startReader() {
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }

    private void handleUpdate(Response response) {
        if (response.type() == ResponseType.SOLD_BILET) {
            Bilet bilet = (Bilet) response.data();
            logger.info("Ticket sold " + bilet);
            try {
                employeeObserver.biletAdded(bilet);
            } catch (MyException e) {
                logger.error("Error handle update: " + e);
            }
        }
    }

    private boolean isUpdate(Response response) {
        return response.type() == ResponseType.SOLD_BILET;
    }

    @Override
    public void addBilet(Bilet bilet) throws MyException {
        Request req = new Request.Builder().type(RequestType.ADD_BILET).data(bilet).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            logger.error("Error adding bilet" + response.data().toString());
            String err = response.data().toString();
            throw new MyException(err);
        } else {
            logger.info("Bilet added");
        }
    }

    @Override
    public Spectacol getSpectacol(Long id) throws MyException {
        Request req = new Request.Builder().type(RequestType.GET_SPECTACOL).data(id).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            logger.error("Error finding spectacol" + response.data().toString());
            String err = response.data().toString();
            throw new MyException(err);
        } else {
            logger.info("Found spectacol");
        }
        return (Spectacol) response.data();
    }
    @Override
    public Iterable<Spectacol> getAllSpectacole() throws MyException {
        Request req = new Request.Builder().type(RequestType.GET_ALL_SPECTACOLE).data(null).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            logger.error("Error getting spectacole" + response.data().toString());
            String err = response.data().toString();
            throw new MyException(err);
        } else {
            logger.info("Got spectacole");
        }
        return (List<Spectacol>) response.data();
    }

    @Override
    public Angajat login(Angajat employee, IObserver employeeObserver) throws MyException {
        initializeConnection();
        Request req = new Request.Builder().type(RequestType.LOGIN).data(employee).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.OK) {
            this.employeeObserver = employeeObserver;
            logger.info("Logged in");
            return (Angajat) response.data();
        }
        if (response.type() == ResponseType.ERROR) {
            logger.error("Error logging in" + response.data().toString());
            String err = response.data().toString();
            closeConnection();
            throw new MyException(err);
        }
        return null;
    }
    public void logout(Angajat employee, IObserver client) throws MyException {
        Request req = new Request.Builder().type(RequestType.LOGOUT).data(employee).build();
        sendRequest(req);
        Response response = readResponse();
        closeConnection();
        if (response.type() == ResponseType.ERROR) {
            logger.error("Error logging out" + response.data().toString());
            String err = response.data().toString();
            throw new MyException(err);
        } else {
            logger.info("Logged out");
        }
    }
    private class ReaderThread implements Runnable {
        public void run() {
            while (!finished) {
                try {
                    Object response = input.readObject();
                    logger.info("response received " + response);
                    if (isUpdate((Response) response)) {
                        handleUpdate((Response) response);
                    } else {
                        try {
                            qresponses.put((Response) response);
                        } catch (InterruptedException e) {
                            logger.error("Queue putting response error: " + e);
                        }
                    }
                } catch (IOException | ClassNotFoundException e) {
                    if (e instanceof SocketException)
                        logger.info("Socket closed: " + e);
                    else
                        logger.error("Reading error: " + e);
                }
            }
        }
    }
}